kanidmd/src/lib/filter.rs

```diff
@@ -137,22 +140,28 @@ enum FilterComp {
     // Not(Box<FilterComp>),
 }
 
-// This is the fully resolved internal representation. Note the lack of Not and selfUUID
-// because these are resolved into And(Pres(class), AndNot(term)) and Eq(uuid, ...).
-// Importantly, we make this accessible to Entry so that it can then match on filters
-// properly.
-#[derive(Debug, Clone, PartialEq, Eq)]
+/// This is the fully resolved internal representation. Note the lack of Not and selfUUID
+/// because these are resolved into And(Pres(class), AndNot(term)) and Eq(uuid, ...).
+/// Importantly, we make this accessible to Entry so that it can then match on filters
+/// internally.
+///
+/// Each filter that has been resolved also has been enriched with metadata about its
+/// index, and index slope. For the purpose of this module, consider slope as a "weight"
+/// where small value - faster index, larger value - slower index. This metadat is extremely
+/// important for the query optimiser to make decisions about how to re-arrange queries
+/// correctly.
+#[derive(Debug, Clone, Eq)]
 pub enum FilterResolved {
```

Comment:
Can you briefly describe what this type is used for? I assume its for matching on something in a db, but I'm confused on the structure.

Reply:
There is a whole rant about how queries work. Each query has a lifecycle basicly. (There are quite a few types that have enforced compile time state machines to guarantee certain behaviour).

If it's from external, it starts as a "ProtoFilter" which is basicly an unsanisited user request.

we then turn that into a "Filter<FilterInvalid>". This is our "internal" structure of an unsanitised query. 

We can transform, alter, modify, or whatever else to the unsanisited query at this point. 

Next we then sanitise this. We process the filter through to "Filter<FilterValid>" which looks through the schema to ensure that every requested attribute, does exist in the schema (so it will be in the db or not). We also check that every request query key is the correct syntax type for the related element. This prevents a denial of service attack that affects most ldap servers (query an attribute that doesn't exist can force a full table scan). 

Once we have a FilterValid we can NOT change the content because any change introduce MAY be invalid and cause issues. 

A FilterValid is a query that is sanitised but it's not yet ready to work in the database. For example, a valid query request is "selfUUID" aka the UUID of the currently authenticated account. 

The next step is FilterValid to FilterValidResolved. This is where we do some internal transforms such as SelfUUID being extracted to an Eq term from the currently authenticated user (this is super important for access controls!). At this point we also *enrich* the query with metadata about the current indexing state of the database (which are the idx slopse, previously a bool of indexed/vs not). 

This metadata is critical for the underlying backend to know if a query component CAN have an index accessed or not. The point of this change changes that metadata from simply true/false to a U8 which allows indication of no-index (0, or in this case, None because of NonZero U8), or that it does have an index (1-255) and what the value of that index is. That's what's used to optimise the query to put "faster terms first".


The final piece of this that you may be missing is the filter threshold - in filter2idl in be/mod.rs, as a filter is being processed, if the partial candidate set falls below this threshold value, we shortcut and return the partial candidate set, and then perform the remaining filter tests on the entry directly.


This is why if you have:

`class=account and name=quinn`
vs
`name=quinn and class=account`

If you did the first query in that order, youd load a huge index, potentially 60,000 values, and then throw away most of it when you hit "name=quinn" which should be length 1. 

But in the other case, you hit name=quinn which is length one, we are below the threshold, so we shortcut return and we just check "does this entry also have class=account?". This avoids us hitting the really large index.

The goal - create queries that are going to get us below that threshold, faster, and if not, create queries that will narrow down more specific elements quicker that may fail so we avoid larger blocks ogf work unless really needed.

Reply:
> There is a whole rant about how queries work. Each query has a lifecycle basicly. (There are quite a few types that have enforced compile time state machines to guarantee certain behaviour).
> 
> If it's from external, it starts as a "ProtoFilter" which is basicly an unsanisited user request.
> 
> we then turn that into a "Filter". This is our "internal" structure of an unsanitised query.
> 
> We can transform, alter, modify, or whatever else to the unsanisited query at this point.

I saw some "optimizing" functions earlier, is this what those were handling?

> 
> Next we then sanitise this. We process the filter through to "Filter" which looks through the schema to ensure that every requested attribute, does exist in the schema (so it will be in the db or not). We also check that every request query key is the correct syntax type for the related element. This prevents a denial of service attack that affects most ldap servers (query an attribute that doesn't exist can force a full table scan).

What would you estimate the complexity of checking for the correct syntax type of a request query key is? O(1)?

> 
> Once we have a FilterValid we can NOT change the content because any change introduce MAY be invalid and cause issues.

How come changes here may be invalid but changes before were okay? Also what kinds of changes were we doing before?

> 
> A FilterValid is a query that is sanitised but it's not yet ready to work in the database. For example, a valid query request is "selfUUID" aka the UUID of the currently authenticated account.
> 
> The next step is FilterValid to FilterValidResolved. This is where we do some internal transforms such as SelfUUID being extracted to an Eq term from the currently authenticated user (this is super important for access controls!). At this point we also _enrich_ the query with metadata about the current indexing state of the database (which are the idx slopse, previously a bool of indexed/vs not).
> 
> This metadata is critical for the underlying backend to know if a query component CAN have an index accessed or not. The point of this change changes that metadata from simply true/false to a U8 which allows indication of no-index (0, or in this case, None because of NonZero U8), or that it does have an index (1-255) and what the value of that index is. That's what's used to optimise the query to put "faster terms first".
> 
> The final piece of this that you may be missing is the filter threshold - in filter2idl in be/mod.rs, as a filter is being processed, if the partial candidate set falls below this threshold value, we shortcut and return the partial candidate set, and then perform the remaining filter tests on the entry directly.
> 
> This is why if you have:
> 
> `class=account and name=quinn`
> vs
> `name=quinn and class=account`
> 
> If you did the first query in that order, youd load a huge index, potentially 60,000 values, and then throw away most of it when you hit "name=quinn" which should be length 1.
> 
> But in the other case, you hit name=quinn which is length one, we are below the threshold, so we shortcut return and we just check "does this entry also have class=account?". This avoids us hitting the really large index.
> 
> The goal - create queries that are going to get us below that threshold, faster, and if not, create queries that will narrow down more specific elements quicker that may fail so we avoid larger blocks ogf work unless really needed.

So essentially, this PR allows us to take a query and sort the attributes from rarest to most common, so we can search in a way that will most rapidly decrease the candidate entries we have to search through?

Reply:
> I saw some "optimizing" functions earlier, is this what those were handling?

Possibly? I'd need to see what you mean. But essential the call to optimise/fast_optimise can only be done on a ValidResolved query because that's the only one with enough metadata. 

> What would you estimate the complexity of checking for the correct syntax type of a request query key is? O(1)?

Big O and algorithmic complexity generally are actually pointless in any real world discussion. They only give you some hand waving theoretical properities about maybe performance - consider binary search tree vs b-tree/b+tree. bst has better algorithmic complexity, but b-tree absolutely wipes the floor with bst in real world. Turns out a whole generation of computer scientists deluded themself into thinking they were mathematicians and decided to literally ignore how computers actually work. Similar hashmaps are O(1) but let me tell you about "hash map performance" sometime. It's never so simple. Once you add threads, it's even more fun.

So I think the complexity is O(who knows). :) 

The more important metric here is about which datastructures, how we access them in the transaction, and how we do these lookups, have we got cache hot components, etc. 

Syntax validation is done here:

https://github.com/kanidm/kanidm/blob/master/kanidmd/src/lib/filter.rs#L599

We recurse through each filter element, and apply the schema attributes syntax validator to it. SchemaTransaction is a readonly in memory hashmap, which given it's shared between threads, copy-on-write, it will be in the S state of your cpu l1/l2/l3 cache (see MESI). The schema elements are tested to ensure they match their expected partial value type. You'll also notice that the attribute type is "AttrString" - this is actually "SmartString" which is an inline string, which prevents an allocation/dereference, which improves cache locality and performance too.

The bigger gain here, could be to investigate doing this with a stack/iteratively rather than recursion, but there is also a question of where are other gains to be found. 

> How come changes here may be invalid but changes before were okay? Also what kinds of changes were we doing before?

Any change could be invalid because even as programmers, we can make mistakes. Trust me, forcing this validation even to internal server code has already caught and prevented multiple bugs. 

https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fwww.nesglobal.net%2Fwp-content%2Fuploads%2FHierarchy-of-Controls-1-1.png&f=1&nofb=1

Examples of an internal change we make is that most queries are wrapped in an exclusion of tombstone/recycled class types so that we can't see anything in the delete paths.

> So essentially, this PR allows us to take a query and sort the attributes from rarest to most common, so we can search in a way that will most rapidly decrease the candidate entries we have to search through?

Correct! And it means that the decision about how to do that sort is based on YOUR data, not just whatever "we" think will be best.